### ISSUE TITLE
Problem with TransformingCandles, InvalidTimeframeException

### ISSUE THREAD

**Issue by hidden1994** [lppkarl/Trady#46]
Today faced problem with converting candles from lower interval(i.e. 15min) to higher interval(i.e. 30m or 1h)

Application perfectly works on my home pc, but refuses to work on another pc.
I am in russia, moscow (+3 UTC), he is in somewhere in canada(-5 UTC).

So i looked into Transform extension function and sibling methods and heres what i found:

Trady/CandleExtension.cs
In method **IsTimeframesValid**
this method works with 2 candles as i understood and for me 

var candleEndTime = periodInstance.NextTimestamp(candles.ElementAt(i).DateTime);

produces right time and in next if condition it doesnt fall through and doesnt produce false result.

however **periodInstance.NextTimestamp** from his pc changes last numbers in timestamps
As i understood in just rounds timestamp

on my pc:
23:29:59 00:00 -> 23:30:00 00:00
and then compared with
23:44:59 00:00

on his pc
23:29:59 00:00 -> 23:30:00 -04:00
and then compared with
23:44:59 00:00

This somehow produces error, i dont know why, but for now i just made my own transform methods where i doesnt use **periodInstance.NextTimestamp** and just do straight compare

**Comment by lppkarl** [comment]
Hi @hidden1994 , to do an indepth test on this issue, would you mind giving us the test code for this issue? Thanks.

**Comment by neonVoice** [comment]
Importer from binance:
```C#
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Binance.Net;
using Binance.Net.Objects;
using Trady.Core;
using Trady.Core.Infrastructure;
using Trady.Core.Period;

namespace BTPIndicatorsWin.General.Trading
{
  internal class BinanceImporter : IImporter
  {
    public async Task<IReadOnlyList<IOhlcv>> ImportAsync(string symbol, DateTime? startTime = null,
      DateTime? endTime = null, PeriodOption period = PeriodOption.Daily,
      CancellationToken token = default(CancellationToken))
    {
      IReadOnlyList<IOhlcv> candlesList;

      if (startTime == null)
        startTime = DateTime.UtcNow.AddHours(-10);

      if (endTime == null)
        endTime = DateTime.UtcNow;

      var interval = ConvertPeriodToKline(period);

      using (var client = new BinanceClient())
      {
        candlesList = await GetCandlesForSymbol(client, symbol, interval,
          startTime.Value, endTime.Value);
      }

      return candlesList;
    }

    private async Task<IReadOnlyList<IOhlcv>> GetCandlesForSymbol(BinanceClient client, string symbol,
      KlineInterval interval, DateTime startTime, DateTime endTime)
    {
      var binanceKlinesTask = await client.GetKlinesAsync(symbol, interval, startTime, endTime);

      var binanceKlines = binanceKlinesTask.Data;

      var candles = new List<IOhlcv>();

      foreach (var binanceKline in binanceKlines)
      {
        var candle = new Candle(binanceKline.CloseTime, binanceKline.Open, binanceKline.High,
          binanceKline.Low, binanceKline.Close, binanceKline.Volume);

        candles.Add(candle);
      }

      return candles;
    }

    private KlineInterval ConvertPeriodToKline(PeriodOption period)
    {
      KlineInterval interval;

      switch (period)
      {
        case PeriodOption.PerSecond:
          interval = KlineInterval.OneMinute; // wrong
          break;
        case PeriodOption.PerMinute:
          interval = KlineInterval.OneMinute;
          break;
        case PeriodOption.Per5Minute:
          interval = KlineInterval.FiveMinutes;
          break;
        case PeriodOption.Per10Minute:
          interval = KlineInterval.FiveteenMinutes; // wrong
          break;
        case PeriodOption.Per15Minute:
          interval = KlineInterval.FiveteenMinutes;
          break;
        case PeriodOption.Per30Minute:
          interval = KlineInterval.ThirtyMinutes;
          break;
        case PeriodOption.Hourly:
          interval = KlineInterval.OneHour;
          break;
        case PeriodOption.BiHourly:
          interval = KlineInterval.TwoHour;
          break;
        case PeriodOption.Daily:
          interval = KlineInterval.OneDay;
          break;
        case PeriodOption.Weekly:
          interval = KlineInterval.OneWeek;
          break;
        case PeriodOption.Monthly:
          interval = KlineInterval.OneMonth;
          break;
        default:
          throw new ArgumentOutOfRangeException(nameof(period), period, null);
      }

      return interval;
    }
  }
}
```

Getting candles history for pair and storing in in "associate array":
```C#
    private static async Task<PairCandles> GetPairCandles(string pairName) {
      var pairCandles = await Importer.ImportAsync(pairName, DateTime.UtcNow.AddHours(-15),
        DateTime.UtcNow, PeriodOption.Per15Minute);

      return new PairCandles(pairCandles, pairName);
    }
```
After returning PairCandles, im immideately start async processing job for it:
```C#
    private static async Task ProcessFilteredPairs() {
      Console.WriteLine("ProcessFilteredPairs -");
      var cnt = 0;

      foreach (var pairName in _filteredPairsList) {
        if (GetOnlyOne) {
          if (cnt++ == 1) {
            break;
          }
        }

        // gets PAIR candles
        var pairCandles = await GetPairCandles(pairName);
        // start analysis on candles asynchronously
        AnalysisTasks.Add(Task.Run(() => FillTradingEnabledLines(pairCandles)));
      }
      // wait for all analysis tasks to end
      await Task.WhenAll(AnalysisTasks);
      Console.WriteLine("ProcessFilteredPairs +");
    }
```

FillTradingenabledLines -> ProcessPairCandles:
```C#
    private static void FillTradingEnabledLines(PairCandles pairCandles) {
      Console.WriteLine("FillTradingEnabledLines - " + pairCandles.Pair);
      var buyProbability = ProcessPairCandles(pairCandles);
      var tradingEnabledString = pairCandles.Pair + Config.LineGeneration.SuffixTradingEnabled;
      tradingEnabledString += buyProbability > Config.Threshold ? "true" : "false";
      TradingEnabledLines.Add(tradingEnabledString);

      ProcessedCandles.Add(pairCandles.Pair, ("BUY", buyProbability));
      Console.WriteLine("FillTradingEnabledLines + " + pairCandles.Pair);
    }
```
ProcessPairCandles -> AnalyzePair:
```C#
    private static double ProcessPairCandles(PairCandles pairCandles) {
      //Console.WriteLine("ProcessPairCandles -");
      double totalBuyProbability = 0;

      foreach (var period in Config.Periods) {
        var buyProbability = AnalyzePair(pairCandles, period);

        if (double.IsNaN(buyProbability)) {
          buyProbability = 0.5;
        }

        var coefficient = PeriodCoefficients[period.Time];
        var multiplied = buyProbability * coefficient;

        //Console.WriteLine($"{pairCandles.Pair} BUY - {buyProbability:0.00} * {coefficient:0.00} = {multiplied:0.00}");
        totalBuyProbability += multiplied;
      }

      // Gets Mean BuyProbability from N Periods (Fast + Slow)
      var newTotal = totalBuyProbability / Config.Periods.Count;

      //Console.WriteLine($"{pairCandles.Pair} BUY - {totalBuyProbability:0.00} / {Config.Periods.Count} = {newTotal:0.00} <- TOTAL");

      //Console.WriteLine("ProcessPairCandles +");
      return newTotal;
    }
```
AnalyzePair -> Extension methods

```C#
    private static double AnalyzePair(PairCandles pairCandles, Config.Period period) {
      //Console.WriteLine("AnalyzePair -");
      var candles = TradyExtensions.ConvertCandlesToTime(pairCandles.Candles, period.Time);
/*
...
*/
      }
```

Extension ConvertCandlesToTime:
```C#
    public static IReadOnlyList<IOhlcv> ConvertCandlesToTime(IReadOnlyList<IOhlcv> inputCandles, string periodTime)
    {
      //Console.WriteLine("ConvertCandlesToTime -");

      var periodOption = GetPeriodFromConfig(periodTime);

      //Console.WriteLine($"periodTime - {periodTime}\tperiodOption - {periodOption}");
      // Supported period: PerSecond, PerMinute, Per15Minutes, Per30Minutes, Hourly, BiHourly, Daily, Weekly, Monthly

      IReadOnlyList<IOhlcv> outputCandles;
      try
      {
        switch (periodOption)
        {
          case PeriodOption.PerSecond:
            outputCandles = Transform<Per15Minute, PerSecond>(inputCandles);
            break;
          case PeriodOption.PerMinute:
            outputCandles = Transform<Per15Minute, PerMinute>(inputCandles);
            break;
          case PeriodOption.Per5Minute:
            outputCandles = Transform<Per15Minute, Per5Minute>(inputCandles);
            break;
          case PeriodOption.Per10Minute:
            outputCandles = Transform<Per15Minute, Per10Minute>(inputCandles);
            break;
          case PeriodOption.Per15Minute:
            outputCandles = Transform<Per15Minute, Per15Minute>(inputCandles);
            break;
          case PeriodOption.Per30Minute:
            outputCandles = Transform<Per15Minute, Per30Minute>(inputCandles);
            break;
          case PeriodOption.Hourly:
            outputCandles = Transform<Per15Minute, Hourly>(inputCandles);
            break;
          case PeriodOption.BiHourly:
            outputCandles = Transform<Per15Minute, BiHourly>(inputCandles);
            break;
          case PeriodOption.Daily:
            outputCandles = Transform<Per15Minute, Daily>(inputCandles);
            break;
          case PeriodOption.Weekly:
            outputCandles = Transform<Per15Minute, Weekly>(inputCandles);
            break;
          case PeriodOption.Monthly:
            outputCandles = Transform<Per15Minute, Monthly>(inputCandles);
            break;
          default:
            outputCandles = inputCandles;
            break;
        }
      }
      catch (Exception e)
      {
        outputCandles = inputCandles;
      }

      //Console.WriteLine("ConvertCandlesToTime +");
      return outputCandles;
    }
```

Here Transform operation is mine with changed code from your library.

**Comment by hidden1994** [comment]
Importer from binance:
```C#
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Binance.Net;
using Binance.Net.Objects;
using Trady.Core;
using Trady.Core.Infrastructure;
using Trady.Core.Period;

namespace BTPIndicatorsWin.General.Trading
{
  internal class BinanceImporter : IImporter
  {
    public async Task<IReadOnlyList<IOhlcv>> ImportAsync(string symbol, DateTime? startTime = null,
      DateTime? endTime = null, PeriodOption period = PeriodOption.Daily,
      CancellationToken token = default(CancellationToken))
    {
      IReadOnlyList<IOhlcv> candlesList;

      if (startTime == null)
        startTime = DateTime.UtcNow.AddHours(-10);

      if (endTime == null)
        endTime = DateTime.UtcNow;

      var interval = ConvertPeriodToKline(period);

      using (var client = new BinanceClient())
      {
        candlesList = await GetCandlesForSymbol(client, symbol, interval,
          startTime.Value, endTime.Value);
      }

      return candlesList;
    }

    private async Task<IReadOnlyList<IOhlcv>> GetCandlesForSymbol(BinanceClient client, string symbol,
      KlineInterval interval, DateTime startTime, DateTime endTime)
    {
      var binanceKlinesTask = await client.GetKlinesAsync(symbol, interval, startTime, endTime);

      var binanceKlines = binanceKlinesTask.Data;

      var candles = new List<IOhlcv>();

      foreach (var binanceKline in binanceKlines)
      {
        var candle = new Candle(binanceKline.CloseTime, binanceKline.Open, binanceKline.High,
          binanceKline.Low, binanceKline.Close, binanceKline.Volume);

        candles.Add(candle);
      }

      return candles;
    }

    private KlineInterval ConvertPeriodToKline(PeriodOption period)
    {
      KlineInterval interval;

      switch (period)
      {
        case PeriodOption.PerSecond:
          interval = KlineInterval.OneMinute; // wrong
          break;
        case PeriodOption.PerMinute:
          interval = KlineInterval.OneMinute;
          break;
        case PeriodOption.Per5Minute:
          interval = KlineInterval.FiveMinutes;
          break;
        case PeriodOption.Per10Minute:
          interval = KlineInterval.FiveteenMinutes; // wrong
          break;
        case PeriodOption.Per15Minute:
          interval = KlineInterval.FiveteenMinutes;
          break;
        case PeriodOption.Per30Minute:
          interval = KlineInterval.ThirtyMinutes;
          break;
        case PeriodOption.Hourly:
          interval = KlineInterval.OneHour;
          break;
        case PeriodOption.BiHourly:
          interval = KlineInterval.TwoHour;
          break;
        case PeriodOption.Daily:
          interval = KlineInterval.OneDay;
          break;
        case PeriodOption.Weekly:
          interval = KlineInterval.OneWeek;
          break;
        case PeriodOption.Monthly:
          interval = KlineInterval.OneMonth;
          break;
        default:
          throw new ArgumentOutOfRangeException(nameof(period), period, null);
      }

      return interval;
    }
  }
}
```

Getting candles history for pair and storing in in "associate array":
```C#
    private static async Task<PairCandles> GetPairCandles(string pairName) {
      var pairCandles = await Importer.ImportAsync(pairName, DateTime.UtcNow.AddHours(-15),
        DateTime.UtcNow, PeriodOption.Per15Minute);

      return new PairCandles(pairCandles, pairName);
    }
```
After returning PairCandles, im immideately start async processing job for it:
```C#
    private static async Task ProcessFilteredPairs() {
      Console.WriteLine("ProcessFilteredPairs -");
      var cnt = 0;

      foreach (var pairName in _filteredPairsList) {
        if (GetOnlyOne) {
          if (cnt++ == 1) {
            break;
          }
        }

        // gets PAIR candles
        var pairCandles = await GetPairCandles(pairName);
        // start analysis on candles asynchronously
        AnalysisTasks.Add(Task.Run(() => FillTradingEnabledLines(pairCandles)));
      }
      // wait for all analysis tasks to end
      await Task.WhenAll(AnalysisTasks);
      Console.WriteLine("ProcessFilteredPairs +");
    }
```

FillTradingenabledLines -> ProcessPairCandles:
```C#
    private static void FillTradingEnabledLines(PairCandles pairCandles) {
      Console.WriteLine("FillTradingEnabledLines - " + pairCandles.Pair);
      var buyProbability = ProcessPairCandles(pairCandles);
      var tradingEnabledString = pairCandles.Pair + Config.LineGeneration.SuffixTradingEnabled;
      tradingEnabledString += buyProbability > Config.Threshold ? "true" : "false";
      TradingEnabledLines.Add(tradingEnabledString);

      ProcessedCandles.Add(pairCandles.Pair, ("BUY", buyProbability));
      Console.WriteLine("FillTradingEnabledLines + " + pairCandles.Pair);
    }
```
ProcessPairCandles -> AnalyzePair:
```C#
    private static double ProcessPairCandles(PairCandles pairCandles) {
      //Console.WriteLine("ProcessPairCandles -");
      double totalBuyProbability = 0;

      foreach (var period in Config.Periods) {
        var buyProbability = AnalyzePair(pairCandles, period);

        if (double.IsNaN(buyProbability)) {
          buyProbability = 0.5;
        }

        var coefficient = PeriodCoefficients[period.Time];
        var multiplied = buyProbability * coefficient;

        //Console.WriteLine($"{pairCandles.Pair} BUY - {buyProbability:0.00} * {coefficient:0.00} = {multiplied:0.00}");
        totalBuyProbability += multiplied;
      }

      // Gets Mean BuyProbability from N Periods (Fast + Slow)
      var newTotal = totalBuyProbability / Config.Periods.Count;

      //Console.WriteLine($"{pairCandles.Pair} BUY - {totalBuyProbability:0.00} / {Config.Periods.Count} = {newTotal:0.00} <- TOTAL");

      //Console.WriteLine("ProcessPairCandles +");
      return newTotal;
    }
```
AnalyzePair -> Extension methods

```C#
    private static double AnalyzePair(PairCandles pairCandles, Config.Period period) {
      //Console.WriteLine("AnalyzePair -");
      var candles = TradyExtensions.ConvertCandlesToTime(pairCandles.Candles, period.Time);
/*
...
*/
      }
```

Extension ConvertCandlesToTime:
```C#
    public static IReadOnlyList<IOhlcv> ConvertCandlesToTime(IReadOnlyList<IOhlcv> inputCandles, string periodTime)
    {
      //Console.WriteLine("ConvertCandlesToTime -");

      var periodOption = GetPeriodFromConfig(periodTime);

      //Console.WriteLine($"periodTime - {periodTime}\tperiodOption - {periodOption}");
      // Supported period: PerSecond, PerMinute, Per15Minutes, Per30Minutes, Hourly, BiHourly, Daily, Weekly, Monthly

      IReadOnlyList<IOhlcv> outputCandles;
      try
      {
        switch (periodOption)
        {
          case PeriodOption.PerSecond:
            outputCandles = Transform<Per15Minute, PerSecond>(inputCandles);
            break;
          case PeriodOption.PerMinute:
            outputCandles = Transform<Per15Minute, PerMinute>(inputCandles);
            break;
          case PeriodOption.Per5Minute:
            outputCandles = Transform<Per15Minute, Per5Minute>(inputCandles);
            break;
          case PeriodOption.Per10Minute:
            outputCandles = Transform<Per15Minute, Per10Minute>(inputCandles);
            break;
          case PeriodOption.Per15Minute:
            outputCandles = Transform<Per15Minute, Per15Minute>(inputCandles);
            break;
          case PeriodOption.Per30Minute:
            outputCandles = Transform<Per15Minute, Per30Minute>(inputCandles);
            break;
          case PeriodOption.Hourly:
            outputCandles = Transform<Per15Minute, Hourly>(inputCandles);
            break;
          case PeriodOption.BiHourly:
            outputCandles = Transform<Per15Minute, BiHourly>(inputCandles);
            break;
          case PeriodOption.Daily:
            outputCandles = Transform<Per15Minute, Daily>(inputCandles);
            break;
          case PeriodOption.Weekly:
            outputCandles = Transform<Per15Minute, Weekly>(inputCandles);
            break;
          case PeriodOption.Monthly:
            outputCandles = Transform<Per15Minute, Monthly>(inputCandles);
            break;
          default:
            outputCandles = inputCandles;
            break;
        }
      }
      catch (Exception e)
      {
        outputCandles = inputCandles;
      }

      //Console.WriteLine("ConvertCandlesToTime +");
      return outputCandles;
    }
```

Here Transform operation is mine with changed code from your library.

**Comment by lppkarl** [comment]
@hidden1994 Thanks for the follow-up. I'll take a look of it 😃

**Comment by codebeaulieu** [comment]
I have a possible fix for this, but I wanted to talk it through to make sure it made sense.

The original IsTimeframesValid method looks like this:
```private static bool IsTimeframesValid<TPeriod>(IEnumerable<IOhlcv> candles, out IOhlcv err)
            where TPeriod : IPeriod
        {
            var periodInstance = Activator.CreateInstance<TPeriod>();
            err = default;
            for (int i = 0; i < candles.Count() - 1; i++)
            {
                var candleEndTime = periodInstance.NextTimestamp(candles.ElementAt(i).DateTime);
                if (candleEndTime > candles.ElementAt(i + 1).DateTime)
                {
                    err = candles.ElementAt(i);
                    return false;
                }
            }
            return true;
        }```

During the check, the offset of the **local machine is applied** to `candleEndTime`, in my case, an offset of -6 hours was applied to the check so `if (candleEndTime > candles.ElementAt(i + 1).DateTime)` always fails. So we need to take the offset that is used in the candles and apply it to `candleEndTime` so we're checking against the same offset.

So, my solution isn't the most elegant, but it works. Hopefully, this will at least start a dialog to lead to a cleaner solution.

```private static bool IsTimeframesValid<TPeriod>(IEnumerable<IOhlcv> candles, out IOhlcv err)
            where TPeriod : IPeriod
        {

            var periodInstance = Activator.CreateInstance<TPeriod>();
            err = default;
            var offset = -1;
            for (int i = 0; i < candles.Count() - 1; i++)
            {
                if (offset == -1)
                    offset = candles.ElementAt(i + 1).DateTime.Offset.Hours;

                var nextTime = periodInstance.NextTimestamp(candles.ElementAt(i).DateTime);
                var candleEndTime = new DateTimeOffset(periodInstance.NextTimestamp(candles.ElementAt(i).DateTime).Date, TimeSpan.FromHours(offset));  
                if (candleEndTime > candles.ElementAt(i + 1).DateTime)
                {
                    err = candles.ElementAt(i);
                    return false;
                }
            }
            return true;
        }```

I'm just storing the local offset in a variable and adding another step that creates a new DateTimeOffset with our offset applied.

This works in my test. Let me know your thoughts. @lppkarl

**Comment by lppkarl** [comment]
@codebeaulieu Thanks for your solution on this issue, i will check in detail in this weekend 👍

**Comment by codebeaulieu** [comment]
@lppkarl Sounds good. Fwiw, I'd be happy to send a PR if this solution works for you.

**Comment by lppkarl** [comment]
@codebeaulieu I have double-checked your code, i think the 1st implementation is pretty good, but i think the block
``` 
if (offset == -1)
    offset = candles.ElementAt(i + 1).DateTime.Offset.Hours;
```
can be moved outside the "if" block and replaced with
```
var offset = candles.Any() ? candles.First().DateTime.Offset.Hours : 0;
```
and may use nextTime directly instead of nextTime.Date in
```
var candleEndTime = new DateTimeOffset(nextTime.Date, TimeSpan.FromHours(offset));
```
as we also allow intraday checking for the candles (The 2nd implementation has the same problem).

Would be appreciated if you could submit a PR on this. Thank you 👍

**Comment by lppkarl** [comment]
should be fixed by #73
